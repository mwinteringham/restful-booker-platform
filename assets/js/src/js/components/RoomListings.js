import React, { useEffect, useState } from 'react';

import RoomListing from './RoomListing.js';
import RoomForm from './RoomForm.js';

import { API } from '../libs/Api.js';

const RoomListings = () => {

	const [rooms, setRooms] = useState([]);

	useEffect(() => {
		updateRooms();
	}, [])

	const updateRooms = () => {
		API.getRoom(setRooms);
	}

	return(
		<div>
			<div className="row">
				<div className="col-sm-1 rowHeader"><p>Room #</p></div>
				<div className="col-sm-2 rowHeader"><p>Type</p></div>
				<div className="col-sm-2 rowHeader"><p>Accessible</p></div>
				<div className="col-sm-1 rowHeader"><p>Price</p></div>
				<div className="col-sm-5 rowHeader"><p>Room details</p></div>
				<div className="col-sm-1"></div>
			</div>
			{rooms.map((room, id) => {
				return <div key={id}><RoomListing details={room} updateRooms={updateRooms} /></div>
			})}
			<RoomForm updateRooms={updateRooms}/>
		</div>
	);
}

export default RoomListings;
