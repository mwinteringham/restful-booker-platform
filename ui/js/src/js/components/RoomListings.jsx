import React from 'react';
import fetch from 'node-fetch';
import RoomListing from './RoomListing.jsx';
import RoomForm from './RoomForm.jsx';
import { API_ROOT } from '../api-config';

export default class RoomListings extends React.Component {

  constructor() {
    super();
		this.state = {
			rooms : [],
			roomNumbers : []
		};
		
		this.updateRooms = this.updateRooms.bind(this);
  }
    
	componentDidMount() {
		this.updateRooms();
	}

	updateRooms() {
		fetch(API_ROOT.room + '/room/')
			.then(res => res.json())
			.then(body => {
				let collectedRoomNumbers = [];
				for(let i =0 ; i < body.rooms.length; i++){
					collectedRoomNumbers.push(body.rooms[i].roomNumber);
				}

				this.setState({rooms : body.rooms, roomNumbers : collectedRoomNumbers});
			});
	}

    render() {
			return(
				<div>
					<div className="row">
						<div className="col-sm-1 rowHeader"><p>Room #</p></div>
						<div className="col-sm-2 rowHeader"><p>Type</p></div>
						<div className="col-sm-2 rowHeader"><p>Accessible</p></div>
						<div className="col-sm-6 rowHeader"><p>Room details</p></div>
						<div className="col-sm-1"></div>
					</div>
					{this.state.rooms.map((room) => {
						return <div key={room.roomid}><RoomListing details={room} updateRooms={this.updateRooms} /></div>
					})}
					<RoomForm roomNumbers={this.state.roomNumbers} updateRooms={this.updateRooms}/>
				</div>
			);
    }
  }