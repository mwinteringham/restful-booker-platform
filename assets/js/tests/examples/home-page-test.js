import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import RoomListings from '../../src/js/components/RoomListings.js';
import nock from 'nock';
import {
  findByText,
  render
} from '@testing-library/react'

// First, we add a mock in for the API that the RoomListings component will load when ready
const roomMock = nock('http://localhost')
                .get('/room/')
                .reply(200, {
                    "rooms": [
                      {
                        "roomid": 1,
                        "roomName": "202",
                        "type": "Single",
                        "accessible": true,
                        "image": "string",
                        "description": "string",
                        "features": [
                          "string"
                        ],
                        "roomPrice": 0
                      }
                    ]
                  });

// We then declare our test by using Jests test function and provide it
// with a name and then an anonymous function which will run our test
test('Rooms list component', async () => {
    // Next we create our component that we want to check. By using
    // react testing libraries' render function and passing it 
    // the React component we want to create, RTL will create a
    // headless DOM and render the React component inside.
    const {asFragment, findByText} = render(
      <BrowserRouter>
        <RoomListings />
      </BrowserRouter>
    );

    await findByText("string")

    // We then compare what the component has done to a previously saved
    // state that lives in the __snapshots__ folder. If anything has changed
    // it will raise an alert.
    expect(asFragment()).toMatchSnapshot();
});

// Check suggestions...
//
// Check creation of components are consistent
// Check state changes in components that use isAuthorised
// Check BookingListings populates with BookingListing components