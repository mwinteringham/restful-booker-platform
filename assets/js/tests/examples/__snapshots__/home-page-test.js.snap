// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms list component 1`] = `
<DocumentFragment>
  <div>
    <div
      class="row"
    >
      <div
        class="col-sm-1 rowHeader"
      >
        <p>
          Room #
        </p>
      </div>
      <div
        class="col-sm-2 rowHeader"
      >
        <p>
          Type
        </p>
      </div>
      <div
        class="col-sm-2 rowHeader"
      >
        <p>
          Accessible
        </p>
      </div>
      <div
        class="col-sm-1 rowHeader"
      >
        <p>
          Price
        </p>
      </div>
      <div
        class="col-sm-5 rowHeader"
      >
        <p>
          Room details
        </p>
      </div>
      <div
        class="col-sm-1"
      />
    </div>
    <div>
      <div
        class="row detail"
        data-testid="roomlisting"
        data-type="room"
        id="room1"
      >
        <div
          class="col-sm-1"
        >
          <p
            id="roomName202"
          >
            202
          </p>
        </div>
        <div
          class="col-sm-2"
        >
          <p
            id="typeSingle"
          >
            Single
          </p>
        </div>
        <div
          class="col-sm-2"
        >
          <p
            id="accessibletrue"
          >
            true
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <p
            id="roomPrice0"
          >
            0
          </p>
        </div>
        <div
          class="col-sm-5"
        >
          <p
            id="detailsstring"
          >
            string
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <span
            class="fa fa-remove roomDelete"
            id="1"
          />
        </div>
      </div>
    </div>
    <div>
      <div
        class="row room-form"
      >
        <div
          class="col-sm-1"
        >
          <input
            class="form-control"
            data-testid="roomName"
            id="roomName"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-sm-2"
        >
          <select
            class="form-control"
            id="type"
          >
            <option
              value="Single"
            >
              Single
            </option>
            <option
              value="Twin"
            >
              Twin
            </option>
            <option
              value="Double"
            >
              Double
            </option>
            <option
              value="Family"
            >
              Family
            </option>
            <option
              value="Suite"
            >
              Suite
            </option>
          </select>
        </div>
        <div
          class="col-sm-2"
        >
          <select
            class="form-control"
            id="accessible"
          >
            <option
              value="false"
            >
              false
            </option>
            <option
              value="true"
            >
              true
            </option>
          </select>
        </div>
        <div
          class="col-sm-1"
        >
          <input
            class="form-control"
            id="roomPrice"
            type="text"
            value=""
          />
        </div>
        <div
          class="col-sm-5"
        >
          <div
            class="row"
          >
            <div
              class="col-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  id="wifiCheckbox"
                  name="featureCheck"
                  type="checkbox"
                  value="WiFi"
                />
                <label
                  class="form-check-label"
                  for="wifiCheckbox"
                >
                  WiFi
                </label>
              </div>
            </div>
            <div
              class="col-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  id="tvCheckbox"
                  name="featureCheck"
                  type="checkbox"
                  value="TV"
                />
                <label
                  class="form-check-label"
                  for="tvCheckbox"
                >
                  TV
                </label>
              </div>
            </div>
            <div
              class="col-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  id="radioCheckbox"
                  name="featureCheck"
                  type="checkbox"
                  value="Radio"
                />
                <label
                  class="form-check-label"
                  for="radioCheckbox"
                >
                  Radio
                </label>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  id="refreshCheckbox"
                  name="featureCheck"
                  type="checkbox"
                  value="Refreshments"
                />
                <label
                  class="form-check-label"
                  for="refreshCheckbox"
                >
                  Refreshments
                </label>
              </div>
            </div>
            <div
              class="col-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  id="safeCheckbox"
                  name="featureCheck"
                  type="checkbox"
                  value="Safe"
                />
                <label
                  class="form-check-label"
                  for="safeCheckbox"
                >
                  Safe
                </label>
              </div>
            </div>
            <div
              class="col-4"
            >
              <div
                class="form-check form-check-inline"
              >
                <input
                  class="form-check-input"
                  id="viewsCheckbox"
                  name="featureCheck"
                  type="checkbox"
                  value="Views"
                />
                <label
                  class="form-check-label"
                  for="viewsCheckbox"
                >
                  Views
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-sm-1"
        >
          <button
            class="btn btn-outline-primary"
            id="createRoom"
            type="submit"
          >
            Create
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
